code/planning/src/planning/preplanner.py

```diff
@@ -0,0 +1,281 @@
+                break
+        # if first target point is turning point
+        if x_target == x_turn and y_target == y_turn:
+            x_target = None
+            y_target = None
+
+        x_turn_follow = data.poses[ind+1].position.x
+        y_turn_follow = data.poses[ind+1].position.y
+
+        self.loginfo(f"x_target = {x_start}")
+        self.loginfo(f"y_target = {y_start}")
+        self.loginfo(f"x_target = {x_turn}")
+        self.loginfo(f"y_target = {y_turn}")
+        self.loginfo(f"x_target = {x_turn_follow}")
+        self.loginfo(f"y_target = {y_turn_follow}")
+        self.loginfo(f"x_target = {x_target}")
+        self.loginfo(f"y_target = {y_target}")
+        # Trajectory for the starting road segment
+        self.odc.initial_road_trajectory(x_start, y_start,
+                                         x_turn, y_turn,
+                                         x_turn_follow, y_turn_follow,
+                                         x_target, y_target,
+                                         0, data.road_options[0])
+        n = len(data.poses)
+
+        # iterating through global route to create trajectory
+        for i in range(1, len(data.road_options)):
+            self.loginfo(f"Preplanner going throug global plan {i+1}/{n}")
+
+            x_target = data.poses[i].position.x
+            y_target = data.poses[i].position.y
+            action = data.road_options[i]
+            # z_target = pose.position.z
+
+            """ roll, pitch, yaw = tf.transformations.euler_from_quaternion(
+                [pose.orientation.x, pose.orientation.y,
+                 pose.orientation.z, pose.orientation.w])
+            action = yaw    # TODO: action should be road_option """
+            # self.loginfo("Yaw {}".format(yaw))
+            # last target reached -> has no follower
+            if i == len(data.road_options) - 1:
+                break
+            x_target_next = data.poses[i+1].position.x
+            y_target_next = data.poses[i+1].position.y
+            self.odc.target_road_trajectory(x_target, y_target,
+                                            x_target_next, y_target_next,
+                                            action)
+            self.loginfo("ROAD ID {}".format(self.odc.road_id))
+            self.loginfo("Action {}".format(action))
+
+        self.odc.target_road_trajectory(x_target, y_target,
+                                        None, None,
+                                        action)
+        self.loginfo("ROAD ID {}".format(self.odc.road_id))
+        self.loginfo("Action {}".format(action))
+        self.loginfo("Trajectory completed!")
+        # trajectory is now stored in the waypoints
+        waypoints = self.odc.waypoints
+        way_x = waypoints[0]
+        way_y = waypoints[1]
+        way_yaw = waypoints[2]
+        way_speed = waypoints[3]
+
+        # Transforming the calculated waypoints into a Path msg
+        # speed is the z coordinate of the path message
```

Comment:
The speed value is the max speed extracted from the openDrive data.
So every point holds the maximum speed permitted on the road segment. 
Because we dont need the z value, i thought it would be easier to publish the max speed in this parameter, to have the trajectory data in one message.